File: app/routes/conversations.py
"""
Empire v7.3 - Conversations API Routes
REST API endpoints for chat conversation and message CRUD operations

Enables persistent chat history in Supabase for the desktop app.
"""

from fastapi import APIRouter, HTTPException, Depends
from typing import Optional
from app.middleware.auth import get_current_user
from app.services.conversation_service import get_conversation_service, ConversationService
from app.models.conversations import (
    CreateConversationRequest,
    UpdateConversationRequest,
    CreateMessageRequest,
    UpdateMessageRequest,
    CreateConversationResponse,
    GetConversationResponse,
    ListConversationsResponse,
    UpdateConversationResponse,
    DeleteConversationResponse,
    CreateMessageResponse,
    ListMessagesResponse,
    UpdateMessageResponse,
    ConversationSortField,
    SortOrder,
)

router = APIRouter(prefix="/api/conversations", tags=["Conversations"])


def get_service() -> ConversationService:
    """Dependency to get the conversation service"""
    return get_conversation_service()


# ==============================================================================
# Conversation Endpoints
# ==============================================================================

@router.post("", response_model=CreateConversationResponse)
async def create_conversation(
    request: CreateConversationRequest,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Create a new conversation"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    return await service.create_conversation(user_id, request)


@router.get("", response_model=ListConversationsResponse)
async def list_conversations(
    project_id: Optional[str] = None,
    is_active: Optional[bool] = None,
    sort_by: ConversationSortField = ConversationSortField.UPDATED_AT,
    sort_order: SortOrder = SortOrder.DESC,
    limit: int = 50,
    offset: int = 0,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """
    List conversations for the current user.

    - **project_id**: Filter by project (optional)
    - **is_active**: Filter by active status (optional)
    - **sort_by**: Field to sort by (created_at, updated_at, last_message_at, title)
    - **sort_order**: Sort order (asc or desc)
    - **limit**: Maximum results to return (default 50)
    - **offset**: Number of results to skip (for pagination)
    """
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    return await service.list_conversations(
        user_id=user_id,
        project_id=project_id,
        is_active=is_active,
        sort_by=sort_by,
        sort_order=sort_order,
        limit=limit,
        offset=offset
    )


@router.get("/{conversation_id}", response_model=GetConversationResponse)
async def get_conversation(
    conversation_id: str,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Get a single conversation by ID"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.get_conversation(conversation_id, user_id)
    if not result.success:
        raise HTTPException(status_code=404, detail=result.error or "Conversation not found")

    return result


@router.put("/{conversation_id}", response_model=UpdateConversationResponse)
async def update_conversation(
    conversation_id: str,
    request: UpdateConversationRequest,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Update a conversation"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.update_conversation(conversation_id, user_id, request)
    if not result.success:
        raise HTTPException(status_code=404, detail=result.error or "Conversation not found")

    return result


@router.delete("/{conversation_id}", response_model=DeleteConversationResponse)
async def delete_conversation(
    conversation_id: str,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Delete a conversation and all its messages"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.delete_conversation(conversation_id, user_id)
    if not result.success:
        raise HTTPException(status_code=404, detail=result.error or "Conversation not found")

    return result


# ==============================================================================
# Message Endpoints
# ==============================================================================

@router.post("/{conversation_id}/messages", response_model=CreateMessageResponse)
async def create_message(
    conversation_id: str,
    request: CreateMessageRequest,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Create a new message in a conversation"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.create_message(conversation_id, user_id, request)
    if not result.success:
        raise HTTPException(status_code=400, detail=result.error or "Failed to create message")

    return result


@router.get("/{conversation_id}/messages", response_model=ListMessagesResponse)
async def list_messages(
    conversation_id: str,
    limit: int = 100,
    offset: int = 0,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """
    List messages in a conversation.

    - **limit**: Maximum messages to return (default 100)
    - **offset**: Number of messages to skip (for pagination)
    """
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.list_messages(conversation_id, user_id, limit, offset)
    if not result.success:
        raise HTTPException(status_code=404, detail="Conversation not found")

    return result


@router.put("/{conversation_id}/messages/{message_id}", response_model=UpdateMessageResponse)
async def update_message(
    conversation_id: str,
    message_id: str,
    request: UpdateMessageRequest,
    user: dict = Depends(get_current_user),
    service: ConversationService = Depends(get_service)
):
    """Update a message"""
    user_id = user.get("sub") or user.get("id")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid user token")

    result = await service.update_message(message_id, conversation_id, user_id, request)
    if not result.success:
        raise HTTPException(status_code=404, detail=result.error or "Message not found")

    return result

File: empire-desktop/src/lib/api/conversations.ts
/**
 * Empire Conversations API
 * CRUD operations for chat conversations stored in Supabase
 *
 * Conversations are persisted in the cloud to survive desktop app updates
 */

import { get, post, del, apiRequest } from './client'
import type { Conversation, Message, Source } from '@/types'

// ============ API Types ============

export type MessageRole = 'user' | 'assistant'

export interface APIConversation {
  id: string
  user_id: string
  project_id: string | null
  title: string
  summary: string | null
  message_count: number
  total_tokens: number
  is_active: boolean
  metadata: Record<string, unknown> | null
  first_message_at: string | null
  last_message_at: string | null
  created_at: string
  updated_at: string
}

export interface APIConversationSummary {
  id: string
  project_id: string | null
  title: string
  message_count: number
  last_message_at: string | null
  created_at: string
  updated_at: string
}

export interface APIMessage {
  id: string
  session_id: string
  message_index: number
  role: MessageRole
  content: string
  sources: Array<Record<string, unknown>> | null
  metadata: Record<string, unknown> | null
  model_name: string | null
  tokens_used: number | null
  processing_time_ms: number | null
  created_at: string
}

export interface CreateConversationRequest {
  title?: string
  project_id?: string
  metadata?: Record<string, unknown>
}

export interface UpdateConversationRequest {
  title?: string
  summary?: string
  is_active?: boolean
  metadata?: Record<string, unknown>
}

export interface CreateMessageRequest {
  role: MessageRole
  content: string
  sources?: Array<Record<string, unknown>>
  metadata?: Record<string, unknown>
  model_name?: string
  tokens_used?: number
  processing_time_ms?: number
}

export interface UpdateMessageRequest {
  content?: string
  sources?: Array<Record<string, unknown>>
  metadata?: Record<string, unknown>
}

// Response types
export interface CreateConversationResponse {
  success: boolean
  conversation?: APIConversation
  message: string
  error?: string
}

export interface GetConversationResponse {
  success: boolean
  conversation?: APIConversation
  error?: string
}

export interface ListConversationsResponse {
  success: boolean
  conversations: APIConversationSummary[]
  total: number
  limit: number
  offset: number
  has_more: boolean
}

export interface UpdateConversationResponse {
  success: boolean
  conversation?: APIConversation
  message: string
  error?: string
}

export interface DeleteConversationResponse {
  success: boolean
  conversation_id?: string
  message: string
  deleted_messages_count: number
  error?: string
}

export interface CreateMessageResponse {
  success: boolean
  message?: APIMessage
  error?: string
}

export interface ListMessagesResponse {
  success: boolean
  messages: APIMessage[]
  total: number
}

export interface UpdateMessageResponse {
  success: boolean
  message?: APIMessage
  error?: string
}

// ============ Helper Functions ============

/**
 * Convert API conversation to client Conversation type
 */
function toClientConversation(apiConv: APIConversation | APIConversationSummary): Conversation {
  return {
    id: apiConv.id,
    projectId: apiConv.project_id ?? undefined,
    title: apiConv.title,
    messageCount: apiConv.message_count,
    lastMessageAt: apiConv.last_message_at ? new Date(apiConv.last_message_at) : undefined,
    createdAt: new Date(apiConv.created_at),
    updatedAt: new Date(apiConv.updated_at),
  }
}

/**
 * Convert API message to client Message type
 */
function toClientMessage(apiMsg: APIMessage): Message {
  return {
    id: apiMsg.id,
    conversationId: apiMsg.session_id,
    role: apiMsg.role,
    content: apiMsg.content,
    sources: apiMsg.sources as Source[] | undefined,
    createdAt: new Date(apiMsg.created_at),
    updatedAt: new Date(apiMsg.created_at), // API doesn't track updated_at for messages
    status: 'complete',
  }
}

// ============ Conversation API Functions ============

/**
 * Get all conversations for the current user
 */
export async function listConversations(options?: {
  projectId?: string
  isActive?: boolean
  sortBy?: 'created_at' | 'updated_at' | 'last_message_at' | 'title'
  sortOrder?: 'asc' | 'desc'
  limit?: number
  offset?: number
}): Promise<Conversation[]> {
  const params: Record<string, string> = {}

  if (options?.projectId) params.project_id = options.projectId
  if (options?.isActive !== undefined) params.is_active = String(options.isActive)
  if (options?.sortBy) params.sort_by = options.sortBy
  if (options?.sortOrder) params.sort_order = options.sortOrder
  if (options?.limit) params.limit = options.limit.toString()
  if (options?.offset) params.offset = options.offset.toString()

  const response = await get<ListConversationsResponse>('/api/conversations', params)

  if (!response.success) {
    throw new Error('Failed to load conversations')
  }

  return response.conversations.map(toClientConversation)
}

/**
 * Get a single conversation by ID
 */
export async function getConversation(id: string): Promise<Conversation> {
  const response = await get<GetConversationResponse>(`/api/conversations/${id}`)

  if (!response.success || !response.conversation) {
    throw new Error(response.error || 'Conversation not found')
  }

  return toClientConversation(response.conversation)
}

/**
 * Create a new conversation
 */
export async function createConversation(
  title: string = 'New Conversation',
  projectId?: string
): Promise<Conversation> {
  const body: CreateConversationRequest = { title }
  if (projectId) body.project_id = projectId

  const response = await post<CreateConversationResponse>('/api/conversations', body)

  if (!response.success || !response.conversation) {
    throw new Error(response.error || 'Failed to create conversation')
  }

  return toClientConversation(response.conversation)
}

/**
 * Update a conversation
 */
export async function updateConversation(
  id: string,
  updates: UpdateConversationRequest
): Promise<Conversation> {
  const response = await apiRequest<UpdateConversationResponse>(`/api/conversations/${id}`, {
    method: 'PUT',
    body: JSON.stringify(updates),
  })

  if (!response.success || !response.conversation) {
    throw new Error(response.error || 'Failed to update conversation')
  }

  return toClientConversation(response.conversation)
}

/**
 * Delete a conversation and all its messages
 */
export async function deleteConversation(id: string): Promise<{ deletedMessagesCount: number }> {
  const response = await del<DeleteConversationResponse>(`/api/conversations/${id}`)

  if (!response.success) {
    throw new Error(response.error || 'Failed to delete conversation')
  }

  return { deletedMessagesCount: response.deleted_messages_count }
}

// ============ Message API Functions ============

/**
 * Get all messages in a conversation
 */
export async function listMessages(
  conversationId: string,
  options?: {
    limit?: number
    offset?: number
  }
): Promise<Message[]> {
  const params: Record<string, string> = {}

  if (options?.limit) params.limit = options.limit.toString()
  if (options?.offset) params.offset = options.offset.toString()

  const response = await get<ListMessagesResponse>(
    `/api/conversations/${conversationId}/messages`,
    params
  )

  if (!response.success) {
    throw new Error('Failed to load messages')
  }

  return response.messages.map(toClientMessage)
}

/**
 * Create a new message in a conversation
 */
export async function createMessage(
  conversationId: string,
  role: MessageRole,
  content: string,
  options?: {
    sources?: Source[]
    modelName?: string
    tokensUsed?: number
    processingTimeMs?: number
  }
): Promise<Message> {
  const body: CreateMessageRequest = {
    role,
    content,
  }

  if (options?.sources) body.sources = options.sources as Array<Record<string, unknown>>
  if (options?.modelName) body.model_name = options.modelName
  if (options?.tokensUsed) body.tokens_used = options.tokensUsed
  if (options?.processingTimeMs) body.processing_time_ms = options.processingTimeMs

  const response = await post<CreateMessageResponse>(
    `/api/conversations/${conversationId}/messages`,
    body
  )

  if (!response.success || !response.message) {
    throw new Error(response.error || 'Failed to create message')
  }

  return toClientMessage(response.message)
}

/**
 * Update a message
 */
export async function updateMessage(
  conversationId: string,
  messageId: string,
  updates: UpdateMessageRequest
): Promise<Message> {
  const response = await apiRequest<UpdateMessageResponse>(
    `/api/conversations/${conversationId}/messages/${messageId}`,
    {
      method: 'PUT',
      body: JSON.stringify(updates),
    }
  )

  if (!response.success || !response.message) {
    throw new Error(response.error || 'Failed to update message')
  }

  return toClientMessage(response.message)
}

/**
 * Sync conversations from API to local cache
 */
export async function syncConversations(projectId?: string): Promise<{
  conversations: Conversation[]
  syncedAt: string
}> {
  const conversations = await listConversations({
    projectId,
    sortBy: 'updated_at',
    sortOrder: 'desc',
  })

  return {
    conversations,
    syncedAt: new Date().toISOString(),
  }
}
